File: pyshmem/sub.py
#!/usr/bin/env python3
"""
Subscriber implementation for the shmem library using NumPy arrays.
"""

import sys
import time
import numpy as np
from numpy.typing import NDArray
from typing import Optional, Tuple, Any
import re

# Import the SMQueue class from our package
from shmem import SMQueue

# Constants
QUEUE_NAME = "/my_queue_example_2"
MESSAGE_SIZE = int(10 * 1024 * 1024)  # 10MB message size
HEADER_SIZE = 64  # Reduced header size

# Global variables
running = True
queue: Optional[SMQueue] = None

# Compile regex pattern for header parsing
HEADER_PATTERN = re.compile(r"Message #(\d+) (\d+)")


def parse_header(header_bytes: NDArray[np.uint8]) -> Tuple[int, int]:
    """
    Parse the header and extract message number and timestamp.

    Args:
        header_bytes: NumPy array containing the header bytes

    Returns:
        Tuple of (message_number, timestamp)

    Raises:
        ValueError: If the header cannot be parsed
    """
    # Convert bytes to string, stopping at the first null byte
    header_str = bytes(header_bytes).split(b"\0")[0].decode("utf-8")

    # Use regex to extract message number and timestamp
    match = HEADER_PATTERN.match(header_str)
    if not match:
        raise ValueError(f"Failed to parse header: {header_str}")

    msg_num = int(match.group(1))
    timestamp = int(match.group(2))

    return msg_num, timestamp


def main() -> None:
    """Main function for the subscriber."""
    global running, queue

    try:
        # Open existing queue
        queue = SMQueue.open(QUEUE_NAME)
        print("Subscriber started. Press Ctrl+C to stop.")
        print(
            f"Element size: {queue.element_size()} bytes, Max elements: {queue.max_elements()}"
        )

        # For calculating running average
        running_avg = 0.0
        alpha = 0.1  # Weight for new samples (0.1 = 10% weight to new sample)
        first_message = True
        message_count = 0
        expected_msg_id = 0  # Track the expected message ID
        
        # Add timing for the entire loop
        loop_start_time = time.time()
        messages_processed = 0
        
        print("Starting message processing loop...")

        while running:
            # Try to pop a message (non-blocking)
            # The profiling is now built into the try_pop_np method
            message: Optional[NDArray[np.uint8]] = queue.pop_np()
            
            if message is not None:
                messages_processed += 1
                # Get current time for latency calculation in milliseconds
                receive_time = time.time() * 1000  # milliseconds since epoch

                # Parse header
                try:
                    msg_num, send_timestamp = parse_header(message[:HEADER_SIZE])

                    # Print basic message info
                    print(f"Received: Message #{msg_num}")
                    print(f"  Send timestamp: {send_timestamp} ms, Receive timestamp: {receive_time} ms")
                    print(f"  Time difference: {receive_time - send_timestamp} ms")

                    # Only calculate transfer time if the message ID matches what we expect
                    if msg_num == expected_msg_id:
                        # Ensure the timestamp is valid
                        if send_timestamp <= 0 or send_timestamp > receive_time:
                            print(f"  Invalid timestamp: {send_timestamp}")
                        else:
                            # Calculate transfer time in milliseconds
                            # Both receive_time and send_timestamp are now in milliseconds
                            transfer_time_ms = receive_time - send_timestamp

                            # Sanity check on transfer time
                            if transfer_time_ms < 0 or transfer_time_ms > 10000:
                                print(
                                    f"  Suspicious transfer time: {transfer_time_ms}ms, ignoring"
                                )
                            else:
                                message_count += 1

                                # Update running average using exponential moving average
                                if first_message:
                                    running_avg = transfer_time_ms
                                    first_message = False
                                else:
                                    running_avg = (
                                        1 - alpha
                                    ) * running_avg + alpha * transfer_time_ms

                                # Print timing information
                                print(f"  Transfer time: {transfer_time_ms:.3f} ms")
                                print(
                                    f"  Running average: {running_avg:.3f} ms (over {message_count} messages)"
                                )
                    else:
                        print(f"  Message ID mismatch. Expected: {expected_msg_id}, Got: {msg_num}")
                    
                    # Update expected message ID for next message
                    expected_msg_id = msg_num + 1

                except ValueError as e:
                    print(f"Error parsing message: {e}")
            else:
                # Queue is empty, wait a bit before trying again
                # Reduce the sleep time to minimize polling delay
                time.sleep(0.001)  # 1 millisecond (was 100 microseconds)
            
            # Calculate and print loop statistics every 10 seconds
            current_time = time.time()
            elapsed_time = current_time - loop_start_time
            if elapsed_time >= 10:
                loop_rate = messages_processed / elapsed_time
                print(f"\nLoop statistics:")
                print(f"  Messages processed: {messages_processed}")
                print(f"  Elapsed time: {elapsed_time:.2f} seconds")
                print(f"  Processing rate: {loop_rate:.2f} messages/second")
                print(f"  Average processing time: {1000/loop_rate if loop_rate > 0 else 0:.2f} ms/message\n")
                
                # Reset statistics
                loop_start_time = current_time
                messages_processed = 0

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: csrc/sub.cpp
#include <signal.h>

#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

#include "shmem.h"

volatile bool running = true;
std::string queue_name = "/my_queue_example_2";
constexpr size_t MESSAGE_SIZE = 10 * 1024 * 1024; // 10MB message size
constexpr size_t HEADER_SIZE = 64;                // Reduced header size

void signal_handler(int signum) { running = false; }

// Fast function to parse the header and extract timestamp
bool parse_header(const char* header, int& msg_num, long long& timestamp) {
    // Skip "Message #"
    const char* p = header;
    while (*p && *p != '#')
        p++;
    if (!*p)
        return false;
    p++; // Skip '#'

    // Parse message number
    char* end;
    msg_num = strtol(p, &end, 10);
    if (p == end)
        return false;

    // Skip to timestamp
    p = end;
    while (*p && (*p == ' ' || *p == '\t'))
        p++;
    if (!*p)
        return false;

    // Parse timestamp
    timestamp = strtoll(p, nullptr, 10);
    return true;
}

int main() {
    // Set up signal handling for clean shutdown
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    try {
        // Open existing queue
        auto queue = shmem::SMQueue::open(queue_name);
        std::cout << "Subscriber started. Press Ctrl+C to stop." << std::endl;
        std::cout << "Element size: " << queue.element_size() << " bytes, Max elements: " << queue.max_elements()
                  << std::endl;

        // For calculating running average - use exponential moving average for efficiency
        double running_avg = 0.0;
        const double alpha = 0.1; // Weight for new samples (0.1 = 10% weight to new sample)
        bool first_message = true;
        int message_count = 0;

        // Set stdout to be unbuffered for more accurate timing
        std::cout.setf(std::ios::unitbuf);

        // Pre-allocate buffer for message - use aligned allocation for better performance
        alignas(64) std::vector<std::byte> buffer(MESSAGE_SIZE);

        // Pre-allocate header buffer
        alignas(64) char header_buffer[HEADER_SIZE + 1]; // +1 for null terminator

        // Use steady_clock for more accurate timing
        using clock = std::chrono::steady_clock;

        while (running) {
            // Try to pop a message (non-blocking)
            if (queue.try_pop(buffer.data())) {
                // Get current time for latency calculation - immediately after receiving
                auto now = clock::now();
                auto receive_time =
                    std::chrono::duration_cast<std::chrono::microseconds>(now.time_since_epoch()).count();

                // Extract header from the message (null-terminate it)
                std::memcpy(header_buffer, buffer.data(), HEADER_SIZE);
                header_buffer[HEADER_SIZE] = '\0';

                // Parse header directly without string conversion
                int msg_num;
                long long send_timestamp;
                if (!parse_header(header_buffer, msg_num, send_timestamp)) {
                    std::cerr << "Failed to parse header: " << header_buffer << std::endl;
                    continue;
                }

                // Ensure the timestamp is valid
                if (send_timestamp <= 0 || send_timestamp > receive_time) {
                    std::cerr << "Invalid timestamp: " << send_timestamp << std::endl;
                    continue;
                }

                // Calculate transfer time in milliseconds
                double transfer_time_ms = (receive_time - send_timestamp) / 1000.0;

                // Sanity check on transfer time
                if (transfer_time_ms < 0 || transfer_time_ms > 10000) {
                    std::cerr << "Suspicious transfer time: " << transfer_time_ms << "ms, ignoring" << std::endl;
                    continue;
                }

                message_count++;

                // Update running average using exponential moving average
                if (first_message) {
                    running_avg = transfer_time_ms;
                    first_message = false;
                } else {
                    running_avg = (1 - alpha) * running_avg + alpha * transfer_time_ms;
                }

                // Print message and timing information
                std::cout << "Received: Message #" << msg_num << std::endl;
                std::cout << "  Transfer time: " << std::fixed << std::setprecision(3) << transfer_time_ms << " ms"
                          << std::endl;
                std::cout << "  Running average: " << std::fixed << std::setprecision(3) << running_avg << " ms (over "
                          << message_count << " messages)" << std::endl;
            } else {
                // Queue is empty, wait a bit before trying again - use a shorter sleep time
                std::this_thread::sleep_for(std::chrono::microseconds(100));
            }
        }

        std::cout << "\nShutting down subscriber..." << std::endl;
        return 0;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
}